analyzers/src/SonarAnalyzer.CSharp/Rules/RedundancyInConstructorDestructorDeclaration.cs

```diff
@@ -89,32 +88,34 @@ private static void CheckConstructorDeclaration(SonarSyntaxNodeReportingContext
             && initializer.ArgumentList != null
             && !initializer.ArgumentList.Arguments.Any();
 
-        private static bool IsConstructorRedundant(ConstructorDeclarationSyntax constructorDeclaration, SemanticModel semanticModel) =>
+        private static bool IsConstructorRedundant(ConstructorDeclarationSyntax constructorDeclaration, Lazy<INamedTypeSymbol> typeSymbol) =>
             constructorDeclaration is { ParameterList.Parameters.Count: 0, Body.Statements.Count: 0 }
-            && (IsSinglePublicConstructor(constructorDeclaration, semanticModel)
+            && !IsStructWithInitializedFieldOrProperty(constructorDeclaration.Parent, typeSymbol)
+            && (IsSinglePublicConstructor(constructorDeclaration, typeSymbol)
                 || constructorDeclaration.Modifiers.Any(SyntaxKind.StaticKeyword));
 
-        private static bool IsSinglePublicConstructor(ConstructorDeclarationSyntax constructorDeclaration, SemanticModel semanticModel) =>
+        private static bool IsStructWithInitializedFieldOrProperty(SyntaxNode parentType, Lazy<INamedTypeSymbol> typeSymbol) =>
+            parentType.IsAnyKind(SyntaxKind.StructDeclaration, SyntaxKindEx.RecordStructDeclaration)
+            && ContainsInitializedFieldOrProperty(typeSymbol.Value);
+
+        private static bool IsSinglePublicConstructor(ConstructorDeclarationSyntax constructorDeclaration, Lazy<INamedTypeSymbol> typeSymbol) =>
             constructorDeclaration.Modifiers.Any(SyntaxKind.PublicKeyword)
             && IsInitializerEmptyOrRedundant(constructorDeclaration.Initializer)
-            && TypeHasExactlyOneConstructor(constructorDeclaration, semanticModel);
+            && TypeHasExactlyOneConstructor(typeSymbol.Value);
 
         private static bool IsInitializerEmptyOrRedundant(ConstructorInitializerSyntax initializer) =>
             initializer is null
             || (initializer.ArgumentList.Arguments.Count == 0
                 && initializer.ThisOrBaseKeyword.IsKind(SyntaxKind.BaseKeyword));
 
-        private static bool TypeHasExactlyOneConstructor(BaseMethodDeclarationSyntax constructorDeclaration, SemanticModel semanticModel)
-        {
-            var symbol = semanticModel.GetDeclaredSymbol(constructorDeclaration);
-            return symbol != null
-                   && symbol.ContainingType
-                            .GetMembers()
-                            .OfType<IMethodSymbol>()
-                            .Count(m => m.MethodKind == MethodKind.Constructor && !m.IsImplicitlyDeclared) == 1;
-        }
+        private static bool TypeHasExactlyOneConstructor(INamedTypeSymbol typeSymbol) =>
```

Comment:
I would suggest to change the signature to
```suggestion
        private static bool TypeHasExactlyOneConstructor(SyntaxNode /* or BaseTypeDeclarationSyntax */ containingTypeDeclaration) =>
```
That way, you do not need to pass the lazy symbol factory.